File: backends/python-fastapi/app/routers/gate.py
"""
Gate ルーターモジュール。

Author: Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
"""
from fastapi import APIRouter, Depends, Query
from app.routers.deps import get_gate_interactor, get_bearer_token
from app.usecase.gate.interactor import GateInteractor
from app.usecase.gate.dto import GateIssueDto, GateVerifyDto

router = APIRouter()


@router.get("/gate/issue")
def issue(
    member: str = Query(...),
    token: str = Depends(get_bearer_token),
    interactor: GateInteractor = Depends(get_gate_interactor),
):
    dto = GateIssueDto(access_token=token, member=member)
    vo = interactor.issue_token(dto)
    return {"token": vo.token}


@router.get("/gate/client/{identifier}/verify")
def verify(
    identifier: str,
    token: str = Query(...),
    interactor: GateInteractor = Depends(get_gate_interactor),
):
    dto = GateVerifyDto(identifier=identifier, token=token)
    vo = interactor.verify(dto)
    return {
        "identifier": vo.identifier,
        "member": vo.member,
        "fingerprint": vo.fingerprint,
        "payload": vo.payload,
    }

File: backends/ts-hono/src/routes/gate.ts
/**
 * Gate ルーターモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
import { Hono } from "hono";
import { badRequest, unauthorized } from "../lib/errors.js";
import { db } from "../db/client.js";
import { DrizzleClientRepository } from "../infrastructure/persistence/drizzleClientRepository.js";
import { DrizzleJwtHistoryRepository } from "../infrastructure/persistence/drizzleJwtHistoryRepository.js";
import { RedisGateRepository } from "../infrastructure/cache/redisGateRepository.js";
import { GateInteractor } from "../usecase/gate/interactor.js";

const app = new Hono();

app.get("/gate/issue", async (c) => {
  const auth = c.req.header("Authorization") ?? "";
  const token = auth.startsWith("Bearer ") ? auth.slice(7) : "";
  if (!token) throw unauthorized("token_required");

  const member = c.req.query("member");
  if (!member) throw badRequest("member_required");

  const uc = new GateInteractor(new DrizzleClientRepository(db), new RedisGateRepository(), new DrizzleJwtHistoryRepository(db));
  const vo = await uc.issueToken(token, member);
  return c.json({ token: vo.token });
});

app.get("/gate/client/:identifier/verify", async (c) => {
  const identifier = c.req.param("identifier");
  const token = c.req.query("token");
  if (!token) throw badRequest("token_required");

  const uc = new GateInteractor(new DrizzleClientRepository(db), new RedisGateRepository());
  const vo = await uc.verify(identifier, token);
  return c.json({ identifier: vo.identifier, member: vo.member, fingerprint: vo.fingerprint, payload: vo.payload });
});

export default app;
